Add FFI functions to get and read a wallets Seed Words

philipr-za opened the pull request:
## Description
This PR adds the FFI functions to get the seed words that represent the seed private key of the key manager that wallet is using. This PR also provides the FFI functions used to restructure the collection of seed words and free the memory when you are done with it.

## How Has This Been Tested?
Test provided that exercises all these functions.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

StriderDM commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -707,6 +709,89 @@ pub unsafe extern "C" fn private_key_from_hex(key: *const c_char, error_out: *mu
+/// # Safety
+/// None
+#[no_mangle]
+pub unsafe extern "C" fn seed_words_get_length(seed_words: *mut TariSeedWords, error_out: *mut c_int) -> c_uint {
```
This should be `seed_words: *const TariSeedWords`, since you want to enforce that this method doesn't change what this pointer is pointing to. We'll have to go back and update const correctness for all the methods.

philipr-za replied:
Ah ok, yes I was just copy-pasta the othe rmethods that do this. Will update

StriderDM commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -707,6 +709,89 @@ pub unsafe extern "C" fn private_key_from_hex(key: *const c_char, error_out: *mu
+#[no_mangle]
+pub unsafe extern "C" fn seed_words_get_at(
+    seed_words: *mut TariSeedWords,
+    position: c_uint,
```
Position can be reduced in size to `position: c_ushort`, will work as is. Could possibly be stated as i8 as well (with int8_t in the header, though that may not work everywhere).

philipr-za replied:
i8 is surely too small. I think c_uint is better here because it lines up with usize on the majority of modern systems.

philipr-za committed:
Add FFI functions to get and read a wallets Seed Words

This PR adds the FFI functions to get the seed words that represent the seed private key of the key manager that wallet is using. This PR also provides the FFI functions used to restructure the collection of seed words and free the memory when you are done with it.

CjS77 reviewed (approved):
LGTM but feel like it would have been simpler in impl and API to return a single space-separated string?
